### ISSUE TITLE
ca65: .macpack doesn't look for files in cc65 share dir

### ISSUE THREAD

**Issue by rofl0r** [cc65/cc65#1726]
```
ca65 -I common -o 01-implied.o 01-implied.a
common/macros.a:28: Error: Cannot open include file 'longbranch.mac': No such file or directory
```
the file is available in /share/cc65/asminc/longbranch.mac
i.e. in one of cc65's own directories.

strace from the above:
```
access("common/longbranch.mac", F_OK)   = -1 ENOENT (No such file or directory)
access("longbranch.mac", F_OK)          = -1 ENOENT (No such file or directory)
access("common/longbranch.mac", F_OK)   = -1 ENOENT (No such file or directory)
access("longbranch.mac", F_OK)          = -1 ENOENT (No such file or directory)
writev(2, [{"", 0}, {"common/macros.a:28: Error: Cannot open include file 'longbranch.mac': No such file or directory\n", 96}], 2common/macros.a:28: Error: Cannot open include file 'longbranch.mac': No such file or directory
) = 96
```

the expected behaviour is that the assembler always looks for files included via .macpack in it's own include dir (if the above search in explicitly specified include dirs fails), much like e.g. gcc search for headers it ships inside its own directory tree, e.g. when you `#include <xmmintrin.h>`

**Comment by spiro-trikaliotis** [comment]
Well, you are doing something differently than I do:
```
$ ca65 --version
ca65 V2.19 - Git 451acb342
$ cat test.a 
        .macpack longbranch
$ ca65 -o test.o test.a
$ 
```
I am using the pre-built package from OBS (https://build.opensuse.org/package/show/home:strik/opencbm) on Debian 11.

**Comment by spiro-trikaliotis** [comment]
I forgot one important line:
```
$ ca65 -I common -o test.o test.a`
$
```

**Comment by spiro-trikaliotis** [comment]
What platform are you using? I see that on Windows, the behaviour is slightly different than on non-Windows.

The default search paths are, in this order:

1. the contents of the environment variable CA65_INC
2. the contents of the subdir asminc/ of the environment variable CC65_HOME
3. on non-Windows: What is defined in the (compilation variable) CA65_INC
4. on Windows: the asminc/ directory of the directory where the ca65.exe resides

So, the platform is important!

**Comment by spiro-trikaliotis** [comment]
Forget it, you told about strace, so this is not Windows...

How did you compile ca65? Did you use a PREFIX? If not, the datadir is hard-coded to the compilation directory (..), cf. src/Makefile:23.

**Comment by rofl0r** [comment]
my buildscript is
```sh
test -z "$butch_prefix" && butch_prefix=/
mk_target() {
make -j$MAKE_THREADS \
  CC="$CC" AR="$CROSS_COMPILE"ar USER_CFLAGS="$optcflags" \
  LDFLAGS="$optldflags -Wl,-rpath-link=$butch_root_dir$butch_prefix/lib" \
  PREFIX="$butch_prefix" DESTDIR="$butch_install_dir" \
  "$@"
}
mk_target all
mk_target install
```

the first line is necessary because cc65 complains that empty prefix isn't valid, when it reality it is, on my system there's no /usr, everything goes into / (i.e. /bin/ca65, /share/cc65/...). so instead i pass PREFIX=/
commit used is 945291fd69d9eed2

**Comment by rofl0r** [comment]
(the empty prefix would be necessary so concatenating / and /bin doesnt result in //bin)

**Comment by mrdudz** [comment]
> the empty prefix would be necessary so concatenating / and /bin doesnt result in //bin

although it shouldnt be a problem, can you try

```test -z "$butch_prefix" && butch_prefix=/.```

allowing an empty prefix would be the proper fix though

**Comment by rofl0r** [comment]
> although it shouldnt be a problem

right, it's mostly a non-issue, only cosmetic, however afaik // has some special meaning when NFS is used (but i don't use that).

> allowing an empty prefix would be the proper fix though

indeed

> can you try
> test -z "$butch_prefix" && butch_prefix=/.

i could, yeah, but i kinda see no point in it. or do you think the original issue reported here (minus the "empty prefix not allowed" thingie) has something to do with ca65 not looking into system include dirs ?

**Comment by mrdudz** [comment]
> or do you think the original issue reported here (minus the "empty prefix not allowed" thingie) has something to do with ca65 not looking into system include dirs ?

I might have found the problem. IMHO the logic applied to PREFIX vs DESTDIR is simply wrong and broken. Eg in libsrc/Makefile there is

```
datadir = $(PREFIX)/share/cc65
```
and then later
```
$(if $(PREFIX),,$(error variable "PREFIX" must be set))
$(INSTALL) -d $(DESTDIR)$(datadir)/$(dir)
$(INSTALL) -m0644 ../$(dir)/*.* $(DESTDIR)$(datadir)/$(dir)
```
That means the libs are installed into $(DESTDIR)/$(PREFIX)/share/cc65

however, in src/Makefile there is

```
datadir := $(if $(PREFIX),$(PREFIX)/share/cc65,$(abspath ..))
CA65_INC = $(datadir)/asminc
```
That means CA65_INC will be $(PREFIX)/share/cc65/asminc

but the same src/Makefile installs like
```
bindir  := $(PREFIX)/bin
	$(INSTALL) -d $(DESTDIR)$(bindir)
	$(INSTALL) ../bin/* $(DESTDIR)$(bindir)
```
so the binaries go to $(DESTDIR)/$(PREFIX)/bin

This seems needlessly complicated (and also broken) - i don't see the point of having bothe DESTDIR and PREFIX at all - it should simply use the PREFIX path.

@oliverschmidt could you please elaborate what the intend of having two variables were and what exactly they were supposed to do?

To me it looks like simply DESTDIR should be eliminated - and everything will start working.

It also looks like "make install" isnt used by anyone, and wasn't tested a lot either :) (Should add a test for this to GHA)

@rofl0r: what is the value of DESTDIR in your script? Does it work when both PREFIX and DESTDIR contain "/." ?

**Comment by spiro-trikaliotis** [comment]
Don't you dare removing $DESTDIR! ;)

Just kidding. For a short explanation, look here, for example: https://www.gnu.org/prep/standards/html_node/DESTDIR.html

I need it for building the Debian and RPM packages. @oliverschmidt introduced it for me.

It is used in package builds so you can use a "make install" and install it to some $DESTDIR as if it were a root (similar to how a chroot works).

Or, the other way round: If I use $DESTDIR, all path references should be so that when I chroot to $DESTDIR, the system will find all the files it needs.


It is not overly complicated and not broken. It is just: No-one but package maintainers (me, the Debian maintainer, the RPM maintainer at other systems) should use $DESTDIR in the first place.

It is used, for example, here: https://build.opensuse.org/package/view_file/home:strik/cc65/debian.rules?expand=1 and here: https://build.opensuse.org/package/view_file/home:strik/cc65/cc65.spec?expand=1

**Comment by mrdudz** [comment]
AAAHH! So when DESTDIR=/foo/bar and PREFIX=/usr then i can make all install and zip up the contents in /foo/bar and what i get can be unpacked to the root of the fs and ends up in /usr ? Is this a standard symbol/option (like "PREFIX")?

Then mmh. Back to square one. I dont know whats going wrong. The question(s) from my previous post still stand though :)

**Comment by spiro-trikaliotis** [comment]
@mrdudz, that's exactly how it is intended.

It is a GNU "standard". The link which I posted as a "short explanation" above  states that it is a GNU extension.

@rofl0r: You are using $DESTDIR. You should not do this, because the behaviour you get is to be expected then. You are shooting yourself in the foot.

What do you want to achieve with $DESTDIR?

**Comment by mrdudz** [comment]
> It is a GNU "standard". The link which I posted as a "short explanation" above states that it is a GNU extension.

OK! I see it now :)

I didnt consider this usecase before - this way it all makes sense :)

**Comment by rofl0r** [comment]
> To me it looks like simply DESTDIR should be eliminated - and everything will start working.

no, it is darely needed for packaging. my distro uses DESTDIR=/opt/packagename , and then symlinks the files in there to /. for example:

```
/ # la /bin/ca65
lrwxrwxrwx    1 root     root            20 Apr 17 14:22 /bin/ca65 -> ../opt/cc65/bin/ca65
/ # 
```
in my distro, removing a package is as simple as rm -rf /opt/package (this will leave you with some dangling symlinks, thus there's a tiny frontend for this task)

**Comment by rofl0r** [comment]
> however, in src/Makefile there is
> datadir := $(if $(PREFIX),$(PREFIX)/share/cc65,$(abspath ..))
> CA65_INC = $(datadir)/asminc

do you think the issue is that the build system doesn't set CA65_INC in my scenario, and there's some code in ca65 like `#ifdef CA65_INC ... include_system_headers();    #else nop(); ...` ?

**Comment by spiro-trikaliotis** [comment]
> do you think the issue is that the build system doesn't set CA65_INC in my scenario

Might be.
Can you output the exact command-line you get when CC65 is called (src/Makefile:150, for example, by adding an echo before the line)?

**Comment by rofl0r** [comment]
it's actually wrong on all layers. this was needed to make it work: https://0x0.st/oTJG.patch

```diff
--- a/src/Makefile
+++ b/src/Makefile
@@ -20,7 +20,7 @@
 .SUFFIXES:
 
 bindir  := $(PREFIX)/bin
-datadir := $(if $(PREFIX),$(PREFIX)/share/cc65,$(abspath ..))
+datadir := $(PREFIX)/share/cc65
 
 CA65_INC = $(datadir)/asminc
 CC65_INC = $(datadir)/include
@@ -62,7 +62,7 @@
 
 CFLAGS += -MMD -MP -O3 -I common \
           -Wall -Wextra -Wno-char-subscripts $(USER_CFLAGS) \
-          -DCA65_INC="$(CA65_INC)" -DCC65_INC="$(CC65_INC)" -DCL65_TGT="$(CL65_TGT)" \
+          -DCA65_INC=\"$(CA65_INC)\" -DCC65_INC=\"$(CC65_INC)\" -DCL65_TGT="$(CL65_TGT)" \
           -DLD65_LIB="$(LD65_LIB)" -DLD65_OBJ="$(LD65_OBJ)" -DLD65_CFG="$(LD65_CFG)" \
           -DBUILD_ID="$(BUILD_ID)"
 
--- a/src/ca65/incpath.c
+++ b/src/ca65/incpath.c
@@ -76,7 +76,7 @@
 
     /* Add some compiled-in search paths if defined at compile time. */
 #if defined(CA65_INC) && !defined(_WIN32)
-    AddSearchPath (IncSearchPath, STRINGIZE (CA65_INC));
+    AddSearchPath (IncSearchPath, CA65_INC);
 #endif
 
     /* Add paths relative to the parent directory of the Windows binary. */
```

while at it, use also `find -name Makefile -exec sed 's;@$(CC);$(CC);' -i {} \;
` so one can ACTUALLY SEE what the build is doing. i don't get where that fetishism for "silent" builds comes from, but it's certainly not practical. can you AT LEAST make it optional ? the usual way is to let the user provide a V=1 variable to make to make it verbose.

implemented like so

```make
if ($(V)=1)
Q=
else
Q=@
endif

$(Q)$(CC) -foo foo.c -c ...
```

**Comment by rofl0r** [comment]
did this actually work anywhere else ? if so then it must be gcc thinking `-DFOO=//bar` means that bar is a comment, and it ends up as `-DFOO=`. anyway -DCA65_INC was properly set after the first hunk, but it ended up empty in the debugger. hunk 2 made it so it was treated as a string and finally entered the code, and hunk 3 makes it so the string wasn't again wrapped as a string.

**Comment by oliverschmidt** [comment]
> i don't get where that fetishism for "silent" builds comes from

* Many / most users overlook warning messages in verbose builds.
* Some environment simply stop logging build output at a certain amount, see e.g. https://github.com/cc65/cc65/commit/3cc3b39c893f0d58d139f5159c8babcc13ab6031.

**Comment by mrdudz** [comment]
> did this actually work anywhere else ? if so then it must be gcc thinking `-DFOO=//bar` means that bar is a comment, and it ends up as `-DFOO=`. anyway -DCA65_INC was properly set after the first hunk, but it ended up empty in the debugger. hunk 2 made it so it was treated as a string and finally entered the code, and hunk 3 makes it so the string wasn't again wrapped as a string.

It'd really help if you provided the requested info so we can reproduce whatever is happening. That patch looks a bit questionable for that matter - spiro can produce working debian packages afterall, using very much the same mechanisms.

**Comment by rofl0r** [comment]
ok. this is the command line without patch:
```
gcc -c -MMD -MP -O3 -I common -Wall -Wextra -Wno-char-subscripts -gdwarf-3 -O0 -g3 -DCA65_INC="//share/cc65/asminc" -DCC65_INC="//share/cc65/include" -DCL65_TGT="//share/cc65/target" -DLD65_LIB="//share/cc65/lib" -DLD65_OBJ="//share/cc65/lib" -DLD65_CFG="//share/cc65/cfg" -DBUILD_ID="N/A" -o ../wrk/ca65/incpath.o ca65/incpath.c
```
the issue here is that the shell removes the `"` for all those strings.
this means to the preprocessor that whatever is behing the `=` for those definitions are C *tokens*, not strings. as such, `//` is treated as a comment, not as a path. additionally, lots of other stuff could go wrong, including when spaces are in the filename. 
the fix is to use `\"` in the makefile, so the preprocessor inserts the passed values as strings, not as tokens, and then remove the stringify macro for those values which miraculously made this work in the past when the passed string happened to be accepted as a regular C token.

> so we can reproduce whatever is happening.

this is actually simple to reproduce: just pass `PREFIX=/` to make, then install to DESTDIR foo, then run the built ca65 binary from foo and set a breakpoint on FinishIncludePaths. and you will see that CA65_INC passed to AddSearchPath is an empty string("").

**Comment by mrdudz** [comment]
I wonder if there is a good way to test these things in a testcase.....

anyhow, i'd like to fix this a bit cleaner and more throughout... like:
a) remove extra slashes from pathes in the makefile so they dont even end up in those pathes
b) escape the quotes in all those pathes, not just those you needed to make it work for you
c) not remove the if expression for empty prefix, because this is actually needed for the intended "in tree" searching

**Comment by oliverschmidt** [comment]
> the issue here is that the shell removes the " for all those strings.
[...]
> those values which miraculously made this work in the past

Maybe the solution to this miracle is that the shell wasn't involved at all:

_Commands are essentially one-line shell scripts. In effect, make grabs each line and passes it to a subshell for execution. In fact, make can optimize this (relatively) expensive fork/exec algorithm if it can guarantee that omitting the shell will not change the behavior of the program. It checks this by scanning each command line for shell special characters, such as wildcard characters and i/o redirection. If none are found, make directly executes the command without passing it to a subshell._
From 'Managing Projects with GNU Make, 3rd - Chapter 5. Commands'

BTW: This very behavior of GNU Make already played a role for this very project - https://github.com/cc65/cc65/commit/c9438ae1a74c7564f98b5c995ccc1a8f212053fd

**Comment by mrdudz** [comment]
> it might be preferable to just always add the prefix-based incpath and addtionally always the in-tree/relative-path based lookup (without detecting it based on PREFIX), i dont think it will hurt having both lookups always active.

Yeah i think that is probably the best thing to do... and remove the special case for windows in the process, ie make it behave exactly the same on all platforms

**Comment by oliverschmidt** [comment]
> and remove the special case for windows in the process

Can you please elaborate, what exactly you refer to with this statement?

**Comment by rofl0r** [comment]
> In fact, make can optimize this (relatively) expensive fork/exec algorithm if it can guarantee that omitting the shell will not change the behavior of the program.

well, that's an optimization, but the behaviour is "as if" passed to shell.
here's an extract from strace:

3040  execve("/bin/gcc", ["gcc", "-c", "-MMD", "-MP", "-O3", "-I", "common", "-Wall", "-Wextra", "-Wno-char-subscripts", "-DCA65_INC=/home/rofl/tmp/cc65/asminc", "-DCC65_INC=/home/rofl/tmp/cc65/include", "-DCL65_TGT=/home/rofl/tmp/cc65/target", "-DLD65_LIB=/home/rofl/tmp/cc65/lib", "-DLD65_OBJ=/home/rofl/tmp/cc65/lib", "-DLD65_CFG=/home/rofl/tmp/cc65/cfg", "-DBUILD_ID=Git d8482fbd5", "-o", "../wrk/common/intptrstack.o", "common/intptrstack.c"], [/* 28 vars */]) = 0
3041  execve("/opt/gcc650/bin/../libexec/gcc/x86_64-unknown-linux-musl/6.5.0/cc1", ["/opt/gcc650/bin/../libexec/gcc/x86_64-unknown-linux-musl/6.5.0/cc1", "-quiet", "-I", "common", "-iprefix", "/opt/gcc650/bin/../lib/gcc/x86_64-unknown-linux-musl/6.5.0/", "-MMD", "../wrk/common/intptrstack.d", "-MP", "-MQ", "../wrk/common/intptrstack.o", "-D", "CA65_INC=/home/rofl/tmp/cc65/asminc", "-D", "CC65_INC=/home/rofl/tmp/cc65/include", "-D", "CL65_TGT=/home/rofl/tmp/cc65/target", "-D", "LD65_LIB=/home/rofl/tmp/cc65/lib", "-D", "LD65_OBJ=/home/rofl/tmp/cc65/lib", "-D", "LD65_CFG=/home/rofl/tmp/cc65/cfg", "-D", "BUILD_ID=Git d8482fbd5", "common/intptrstack.c", "-quiet", "-dumpbase", "intptrstack.c", "-mtune=generic", "-march=x86-64", "-auxbase-strip", "../wrk/common/intptrstack.o", "-O3", "-Wall", "-Wextra", "-Wno-char-subscripts", "-o", "/tmp/ccmpCOEb.s"], [/* 31 vars */] <unfinished ...>

you can see, the paths are passed without `"`

**Comment by oliverschmidt** [comment]
Maybe I have missed that - at least from my perspective interesting - point: Is it already clarified, why so many cc65 users were able to successfully use it in the past 8 years?

**Comment by rofl0r** [comment]
> why so many cc65 users were able to successfully use it in the past 8 years?

probably because `#ifndef _WIN32`, or they just used distro packages, or they worked around using the env variable CA65_INC and didnt file an issue, or this was changed recently, or they all use `/usr/local` for prefix, or, or, or ...

**Comment by oliverschmidt** [comment]
Me: "Can you please elaborate, what exactly you refer to with this statement?"

You: "The search order spiro listed above. It should be perfectly doable to do it the same on all systems without breaking anything."

Me: Looking up that list...

> the contents of the environment variable CA65_INC
> the contents of the subdir asminc/ of the environment variable CC65_HOME
> on non-Windows: What is defined in the (compilation variable) CA65_INC
> on Windows: the asminc/ directory of the directory where the ca65.exe resides

Me coming to the comclusion that you are about to remove `on Windows: the asminc/ directory of the directory where the ca65.exe resides`

Where did we loose each other?

**Comment by mrdudz** [comment]
> Where did we loose each other?

You interpret "doing it the same everywhere" as "remove the windows specific things". 

But instead what we could do is doing all 4 steps, always. It wont do any harm if some point at invalid pathes.

**Comment by oliverschmidt** [comment]
> You interpret "doing it the same everywhere" as "remove the windows specific things".

I'd still consider this a valid interpretation. But I'm more than happy to learn that it was wrong :-)

I can by the way very well imagine that certain types of invalid paths checked by Windows for each and every (e.g. include) file might degrade the performance significantly.

**Comment by mrdudz** [comment]
> I can by the way very well imagine that certain types of invalid paths checked by Windows for each and every (e.g. include) file might degrade the performance significantly.

By the time this becomes a problem, its time to replace your 386 :)

**Comment by spiro-trikaliotis** [comment]
You all have lost me in your discussions.

@rofl0r: What I don't get: Why can't you just perform a make with prefix="$butch_install_dir", and leave $DESTDIR alone? Does it have to do with the way your distro creates the symlinks?

Which distro is it, btw?

**Comment by oliverschmidt** [comment]
> By the time this becomes a problem, its time to replace your 386 :)

Really? This was no joke at all. Timeouts for e.g. non-existing drives may be several seconds - and the non-existence isn't necessarily cached, if e.g. a network path was mapped to that drive.

But you seem to know everything about Windows, so it's better I go back in the whole I for some reason decided to come out of today ;-)

**Comment by mrdudz** [comment]
> Timeouts for e.g. non-existing drives may be several seconds

Could you tell how any of those pathes would point to non existing drives please?

**Comment by rofl0r** [comment]
> Which distro is it, btw?

the distro is sabotage linux. i linked above to the sabotage linux blog, and a commit in it.

> Why can't you just perform a make with prefix="$butch_install_dir", and leave $DESTDIR alone? Does it have to do with the way your distro creates the symlinks?

you mean abusing prefix as DESTDIR+prefix ? i unfortunately had to this for a few misbehaving packages which don't provide the DESTDIR functionality, but it often leads to issues. the binaries then have the DESTDIR hardcoded, but they shouldnt be aware of the actual filesystem location wehre they're packaged, but only their internal aka prefix-ed filesystem structure. one such issue is for example that python tries to install libraries into /opt/python/lib rather than /lib, which causes that they're not found because /opt/python's contents are only linked once (at the time python package is installed) into the root filesystem.

**Comment by spiro-trikaliotis** [comment]
> the distro is sabotage linux.

Ok, I did not know that distro, so I did not recognize "sabotage" as a distro name. I thought it was just a "funny" name for your blog or so. ;)

> you mean abusing prefix as DESTDIR+prefix ? [...]

Ok, now, I think I halfway understand what you are doing and what your issues are. And now, I also understand why the PR you wrote might fix your issue.

Just to get it right: Your distro looks into EVERY file in your "package directory" and creates symlinks into it? It does this for every single line?
What if you just "rm" the package directory: How are the symlinks cleaned up? Or are they just left dangling?

**Comment by rofl0r** [comment]
> Just to get it right: Your distro looks into EVERY file in your "package directory" and creates symlinks into it?

right, it does this for every file in the DESTDIR (/opt/packagename) of the package.

> What if you just "rm" the package directory: How are the symlinks cleaned up? Or are they just left dangling?

the "butch rm" command first calls "butch unlink" on the package which removes the symlinks, then rm -rf's /opt/packagename:
https://github.com/sabotage-linux/sabotage/blob/master/KEEP/bin/butch-rm#L37

**Comment by mrdudz** [comment]
OK, so, with #1731 merged the next thing would be tackling the special case handling for empty PREFIX in the makefile. This seems less trivial than i thought, unfortunately. There is no good (reliable) way to detect "Windows" automatically in the makefile, is there?

**Comment by polluks** [comment]
- Cygwin
it's COMSPEC says
https://www.oreilly.com/library/view/managing-projects-with/0596006101/ch07.html
- MSYS2
`ifeq ($(shell uname -o),Msys)`

**Comment by mrdudz** [comment]
To reiterate, the challange is to find something that a) detects "we are building for windows" and b) yields the same result regardless of the shell or even cross assembling environment.

This seems to work:
- if we are running on windows, the OS variable gives the same result in cmd.exe and msys2 (and likely any other shell environment) Here the only question is what this OS variable is on legacy systems (before XP) and if we should even bother with them.
- we can check for a msys compiler using gcc -dumpmachine
```
ifdef OS
$(info OS is defined)
ifeq ($OS,Windows_NT)
$(info Windows >= XP detected)
endif
else
ifneq (,$(findstring mingw,$(shell gcc -dumpmachine 2>$(NULLDEV))))
$(info this is mingw)
endif
endif
```

**Comment by spiro-trikaliotis** [comment]
In src/Makefile:71, the build system determines if the suffix exe is to be used (
```
ifdef CMD_EXE
  EXE_SUFFIX=.exe
endif

ifdef CROSS_COMPILE
  EXE_SUFFIX=.exe
endif
```
Shouldn't this be exactly the case(s) when Windows is built?

**Comment by mrdudz** [comment]
Mh the more i think about it, the more i think any kind of autodetection is flawed in one way or another - and we should also have config variables to override both the detected host and target system (like HOST and TARGET - i dont know if there is a standard for this in GNU makefiles, if there is, we should use that of course)

**Comment by rofl0r** [comment]
> Mh the more i think about it, the more i think any kind of autodetection is flawed in one way or another

this raises the question why you need/want windows detection in the makefile in the first place. isn't it sufficient to use `#ifdef WIN32` as is currently done inside the C source ?

**Comment by rofl0r** [comment]
btw, i noticed that cycgwin through some magic automatically appends the .exe suffix to executable makefile targets; not certain about mingw.

**Comment by mrdudz** [comment]
> this raises the question why you need/want windows detection in the makefile in the first place

to get rid of this "empty PREFIX" special case on non windows

> cygwin through some magic automatically appends the .exe suffix to executable makefile targets; not certain about mingw.

mingw does the same - iirc the problem is that this way make will not "see" the .exe files and then build (at least the last linker step) them again and again on subsequent runs even if not required

**Comment by spiro-trikaliotis** [comment]
Isn't CROSS_COMPILE a way to override auto-detection? At least, I cannot find it defined anywhere inside of the cc65 build system, it is used only in "ifdef CROSS_COMPILE" and "CC = $(CROSS_COMPILE)gcc" (and AR) lines.

**Comment by mrdudz** [comment]
> Isn't CROSS_COMPILE a way to override auto-detection?

Not really. It still implies windows. Right now it doesnt mean "we are cross compiling" but "we are crosscompiling for a windows target". Thats exactly what i want to get rid of - the weird assumptions made from totally unrelated things. Neither "PREFIX" nor "CROSS_COMPILE" are related to "are we compiling for windows". We shouldnt have to define unrelated things to magic values to achive something :)

**Comment by mrdudz** [comment]
To clearify: consider the (i know somewhat theoretical) case of compiling binaries for use on a raspi linux system on windows using arm-gcc with cmd.exe. Right now CROSS_COMPILE=arm-elf- would only work half right, it might produce binaries with the right compiler - but it would add .exe suffix

HOWEVER - i think we are loosing focus of this report now :)

[rofl0r](https://github.com/rofl0r) is it working for you now? If so this issue can be closed IMHO and i will add the other things to the todo list (its not very urgent to fix and somewhat unrelated to the issue here)

**Comment by oliverschmidt** [comment]
BTW: https://github.com/cc65/cc65/issues/1303 ;-))

**Comment by mrdudz** [comment]
closed via #1735

**Comment by oliverschmidt** [comment]
Hi,

@Spiro: Thanks for pointing that $DESTDIR aspect out so quickly, meaning
that I hadn't to :-)

Regards,
Oliver
